lib/services/access_request.go

```diff
@@ -1387,17 +1417,40 @@ func (m *RequestValidator) truncateTTL(ctx context.Context, identity tlsca.Ident
 // GetRequestableRoles gets the list of all existent roles which the user is
 // able to request.  This operation is expensive since it loads all existent
 // roles in order to determine the role list.  Prefer calling CanRequestRole
-// when checking against a known role list.
-func (m *RequestValidator) GetRequestableRoles() ([]string, error) {
-	allRoles, err := m.getter.GetRoles(context.TODO())
+// when checking against a known role list. If resource IDs or a login hint
+// are provided, roles will be filtered to only include those that would
+// allow access to the given resource with the given login.
+func (m *RequestValidator) GetRequestableRoles(ctx context.Context, resourceIDs []types.ResourceID, loginHint string) ([]string, error) {
+	allRoles, err := m.getter.GetRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resources, err := m.getUnderlyingResourcesByResourceIDs(ctx, resourceIDs)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	var expanded []string
 	for _, role := range allRoles {
-		if n := role.GetName(); !slices.Contains(m.userState.GetRoles(), n) && m.CanRequestRole(n) {
-			// user does not currently hold this role, and is allowed to request it.
+		n := role.GetName()
+		if slices.Contains(m.userState.GetRoles(), n) || !m.CanRequestRole(n) {
+			continue
+		}
+
+		roleAllowsAccess := true
+		for _, resource := range resources {
+			access, err := m.roleAllowsResource(ctx, role, resource, loginHint)
+			if err != nil {
+				return nil, trace.Wrap(err)
+			}
+			if !access {
+				roleAllowsAccess = false
+			}
+		}
+
+		// user does not currently hold this role, and is allowed to request it.
+		if roleAllowsAccess {
 			expanded = append(expanded, n)
 		}
 	}
```

Comment:
I don't think the comment about leaking the existence of nodes the user isn't supposed to know about was properly resolved. Consider this example:

```yaml
kind: user
metadata:
  name: tester
spec:
  roles: [role-requester]
---
kind: role
metadata:
  name: role-requester
spec:
  allow:
    request:
      roles: [access, auditor]
```
If this user calls `GetAccessCapabilities` with a resource ID, I believe they will get a different response based on whether the resource exists or doesn't exist, despite the fact they shouldn't be allowed to list the resource since they don't have any `search_as_roles`.

If it exists, the `RequestableRoles` in the result will be `[access]`. If it doesn't exist, the `RequestableRoles` in the result will be `[access, auditor]`.

Let me know if this makes sense, and if you need help resolving it. I think at some point we need to test whether the user can list the resource with their expanded roleset including their `search_as_roles`.

Reply:
I've updated the code to filter the resource IDs by whether the user can access them, as well as added tests. Let me know if the new version works, or if something still needs to change.